### ISSUE TITLE
Is null a vaild schema type or not ?

### ISSUE THREAD

**Issue by Gozala** [ipld/ipld#266]
I'm bit confused about `null`, in the schema schema it's not defined as a kind

https://github.com/ipld/ipld/blob/353baf885adebb93191cbe1f7be34f0517e20bbd/specs/schemas/schema-schema.ipldsch#L145-L154

Nor a type 

https://github.com/ipld/ipld/blob/353baf885adebb93191cbe1f7be34f0517e20bbd/specs/schemas/schema-schema.ipldsch#L115-L129

Yet there is test that seems to assume `null` is a kind

https://github.com/ipld/ipld/blob/353baf885adebb93191cbe1f7be34f0517e20bbd/specs/schemas/tests/null.yml-old#L1-L10

So which one is correct and which on isn't ?

It also appears that type defined in prelude `type Null unit representation null` does not compile with js-toolchain, although not sure if it's a bug in js-toolchain or perhaps prelude is out of date ?

**Comment by rvagg** [comment]
Yeah, this is a tricky one. You can see that I've stumbled on this myself here: https://github.com/ipld/js-ipld-schema/blob/9677e35c7d48feb789d6e09f5a7a7d3331c3d3b8/schema-schema.ts#L29

In my opinion it _should_ be defined as a typekind. It should be usable everywhere a scalar is usable. But it has a cardinality of 1, unlike everything else.

But, I think the reason Eric left it out are twofold:

1. `nullable` is a thing and _almost_ everywhere you want a `null` you should be able to use a `nullable`. Unfortunately it's not 100% of the time.
2. `unit` _should_ solve the rest of the % not covered by `nullable` and it has the benefit of having different representations. The cases where you're not doing a `nullable` field, you're using it either as a placeholder or a signal entry.

`type Null unit representation null` should be the appropriate replacement with the current Schema spec.

Regarding the JS tooling: you're correct that it doesn't compile, `unit` is one of the (maybe the) most recent changes to the the schema language and it just hasn't caught up. I don't believe there's even cursory support for it in Go. The closest thing to "support", aside from being in the schema-schema, is that it's in the TypeScript definitions in js-ipld-schema!

So it's a feature that's just waiting for someone (like you!) to show up with a usecase where it's really annoying to not have it and then implement it. Would you like to have a go? I believe that the [parser in js-ipld-schema](https://github.com/ipld/js-ipld-schema/blob/master/ipld-schema.pegjs) should be fairly clear, but I've not had anyone else contribute to it so I may be wrong about that! I'd love feedback on the process of adding features.

**Comment by rvagg** [comment]
Oh, and re the `ipld/specs/schemas/tests/null.yml-old` test .. I think that one actually works in JS? Maybe. At least it used to. All those fixtures come out of my JS implementation which allowed `null` as a typekind. I reckon I was stubborn enough to keep that in there but I haven't checked. Any of those tests with `-old` are not currently used by the Go test suite and may not even be used by the JS suite. So don't take them as a source of truth. They're a source of my personal historical truth which may or may not reflect anyone else's, or even my own present truth!

**Comment by Gozala** [comment]
> Oh, and re the `ipld/specs/schemas/tests/null.yml-old` test .. I think that one actually works in JS? Maybe. At least it used to.

Yeah this works

But the one in prelude does not

https://observablehq.com/d/9cbd647974be5067

**Comment by Gozala** [comment]
> So it's a feature that's just waiting for someone (like you!) to show up with a usecase where it's really annoying to not have it and then implement it. Would you like to have a go?

I have side project around IPLD schemas that I'm hacking on spare time, which I'd love to share once I get around to getting it usable state which attempts to:

1. Add support for generics per https://github.com/ipld/ipld/issues/248
2. Support inline type defs as per https://github.com/ipld/ipld/issues/262
3. Notion of functions https://github.com/ipld/ipld/issues/263 
4. Removes named types in favor of CID based addressing (This is my main gripe with schemas today)
5. Uses lisp notation, because custom syntax doesn't seem to add any value just overhead IMO (anyone else can add alternative syntax if they wish).
6. Plans to support compiling down to the standard schema

I'm not sure where I'll land with that attempt, but I'm sure it's good field study that my result in retrofitting some ideas into current tool chains.

**Comment by Gozala** [comment]
For what it's worth I end up defining Nullables as `type Nullable<T> union { | Null null | Value T } representation kinded` which in structs, gets normalized into just nullable filed with type `T`.

While I do appreciate what Erik was trying to do there, yet I find it ended up in strange place where `null` is neither first class thing nor it is a field modifier. It's kind of tries to be later, yet fails and consequently ends up been neither. Here is the illustration of where it leaks:

```ipldsch
type Message struct {
  message nullable NullableString
}

type NullableString union {
  Null null
  String string
} representation kinded
```

P.S.: I kind of wish null or nullable was not even a think apart from representation strategy of `unit`

**Comment by Gozala** [comment]
> In my opinion it _should_ be defined as a typekind. It should be usable everywhere a scalar is usable. But it has a cardinality of 1, unlike everything else.

Doesn't int has cardinality of 1 as well ? Or do you meant that `nullable` would not affect cardinality of the `null` ?

**Comment by warpfork** [comment]
Chiming in here just to say :100: to 

> I kind of wish null or nullable was not even a thing apart from representation strategy of unit

Yes.  That's absolutely what we should've done.

Null should most definitely have simply been in the prelude as `type Null unit representation null`.

The `nullable` keyword seems nice as sugar, but probably should've been reduced to... sugar.  The implementation of a lot of logic on schemas would be drastically simplified if we desugarred nullable fields to creating (or adding to) a union type with `Null` as a member.
